From 7f3062ccad1dcf69a3235b338e07a589ffa290e9 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 14 Jun 2023 11:23:17 +0200
Subject: [PATCH 0474/1695] clk: stm32mp25: add low-power support for RIF
 configuration

RCC  RIF configuration is lost during Standby 1/2 low power sequences.

Add a power management sequence to handle the preservation of the RCC
RIF configuration.
stm32_rcc_rif_pm_suspend() saves the actual RCC RIF configurations.
stm32_rcc_rif_pm_resume() restores the configuration saved in
stm32_rcc_rif_pm_suspend().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8d2ab80221d2db99890c30cf1f538460227adca6 clk: stm32mp25: add low-power support for RIF configuration]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I7b66ad08e1db3d53dd183e4bb6088153c70c5664
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/319316
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 59 +++++++++++++++++++++++++++++++-
 1 file changed, 58 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 351f626ee..6124f15f2 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -14,6 +14,7 @@
 #include <dt-bindings/clock/stm32mp25-clksrc.h>
 #include <kernel/boot.h>
 #include <kernel/delay.h>
+#include <kernel/pm.h>
 #include <initcall.h>
 #include <io.h>
 #include <kernel/dt.h>
@@ -3712,6 +3713,56 @@ end:
 	return TEE_SUCCESS;
 }
 
+static TEE_Result stm32_rcc_rif_pm_resume(void)
+{
+	return apply_rcc_rif_config(true);
+}
+
+static TEE_Result stm32_rcc_rif_pm_suspend(void)
+{
+	struct stm32_clk_platdata *pdata = &stm32mp25_clock_pdata;
+	unsigned int i = 0;
+
+	for (i = 0; i < RCC_NB_RIF_RES; i++) {
+		pdata->conf_data.cid_confs[i] = io_read32(pdata->rcc_base +
+							  RCC_CIDCFGR(i));
+	}
+
+	for (i = 0; i < RCC_NB_CONFS; i++) {
+		pdata->conf_data.priv_conf[i] = io_read32(pdata->rcc_base +
+							  RCC_PRIVCFGR(i));
+		pdata->conf_data.sec_conf[i] = io_read32(pdata->rcc_base +
+							 RCC_SECCFGR(i));
+		pdata->conf_data.lock_conf[i] = io_read32(pdata->rcc_base +
+							  RCC_RCFGLOCKR(i));
+		pdata->conf_data.access_mask[i] = GENMASK_32(31, 0);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result
+stm32_rcc_rif_pm(enum pm_op op, unsigned int pm_hint,
+		 const struct pm_callback_handle *pm_handle __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	bool is_tdcid = false;
+
+	res = stm32_rifsc_check_tdcid(&is_tdcid);
+	if (res)
+		return res;
+
+	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT) || !is_tdcid)
+		return TEE_SUCCESS;
+
+	if (op == PM_OP_RESUME)
+		res = stm32_rcc_rif_pm_resume();
+	else
+		res = stm32_rcc_rif_pm_suspend();
+
+	return res;
+}
+
 static TEE_Result rcc_rif_config(void)
 {
 	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
@@ -3721,7 +3772,13 @@ static TEE_Result rcc_rif_config(void)
 	if (res)
 		return res;
 
-	return apply_rcc_rif_config(is_tdcid);
+	res = apply_rcc_rif_config(is_tdcid);
+	if (res)
+		panic();
+
+	register_pm_core_service_cb(stm32_rcc_rif_pm, NULL, "stm32-rcc-rif");
+
+	return TEE_SUCCESS;
 }
 
 driver_init_late(rcc_rif_config);
-- 
2.47.2

